commands/command_filter.go

```diff
@@ -0,0 +1,215 @@
+			Exit("Files don't match:\n%s\n%s", mediafile, tmpfile)
+		}
+		Debug("%s exists", mediafile)
+	} else {
+		if err := os.Rename(tmpfile, mediafile); err != nil {
+			Panic(err, "Unable to move %s to %s\n", tmpfile, mediafile)
+		}
+
+		Debug("Writing %s", mediafile)
+	}
+
+	_, err = cleaned.Pointer.Encode(to)
+	return err
+}
+
+func smudge(to io.Writer, reader io.Reader, filename string) error {
+	var pbuf bytes.Buffer
+	reader = io.TeeReader(reader, &pbuf)
+
+	ptr, err := lfs.DecodePointer(reader)
+	if err != nil {
+		// If we tried to decode a pointer out of the data given to us,
+		// and the file was _empty_, write out an empty file in
+		// response. This occurs because when the clean filter
+		// encounters an empty file, and writes out an empty file,
+		// instead of a pointer.
+		//
+		// TODO(taylor): figure out if there is more data on the reader,
+		// and buffer that as well.
+		if len(pbuf.Bytes()) == 0 {
+			if _, cerr := io.Copy(to, &pbuf); cerr != nil {
+				Panic(cerr, "Error writing data to stdout:")
+			}
+			return nil
+		}
+
+		return err
+	}
+
+	lfs.LinkOrCopyFromReference(ptr.Oid, ptr.Size)
+
+	cb, file, err := lfs.CopyCallbackFile("smudge", filename, 1, 1)
+	if err != nil {
+		Error(err.Error())
+	}
+
+	cfg := config.Config
+	download := lfs.FilenamePassesIncludeExcludeFilter(filename, cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
+
+	if filterSmudgeSkip || cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false) {
+		download = false
+	}
+
+	err = ptr.Smudge(to, filename, download, TransferManifest(), cb)
+	if file != nil {
+		file.Close()
+	}
+
+	if err != nil {
+		// Download declined error is ok to skip if we weren't requesting download
+		if !(errors.IsDownloadDeclinedError(err) && !download) {
+			LoggedError(err, "Error downloading object: %s (%s)", filename, ptr.Oid)
+			if !cfg.SkipDownloadErrors() {
+				// TODO: What to do best here?
+				os.Exit(2)
+			}
+		}
+
+		_, err = ptr.Encode(to)
+		return err
+	}
+
+	return nil
+}
+
+func filterCommand(cmd *cobra.Command, args []string) {
+	requireStdin("This command should be run by the Git filter process")
+	lfs.InstallHooks(false)
+
+	s := git.NewObjectScanner(os.Stdin, os.Stdout)
+
+	if err := s.Init(); err != nil {
+		ExitWithError(err)
+	}
+	if err := s.NegotiateCapabilities(); err != nil {
+		ExitWithError(err)
+	}
+
+Scan:
+	for s.Scan() {
+		var err error
+		var w io.Writer
+
+		req := s.Request()
+		if req == nil {
+			break
+		}
+		s.WriteStatus("success")
+
+		switch req.Header["command"] {
+		case "clean":
+			w = git.NewPacketWriter(os.Stdout, cleanFilterBufferCapacity)
+			err = clean(w, req.Payload, req.Header["pathname"])
+		case "smudge":
+			w = git.NewPacketWriter(os.Stdout, smudgeFilterBufferCapacity)
+			err = smudge(w, req.Payload, req.Header["pathname"])
+		default:
+			fmt.Errorf("Unknown command %s", cmd)
+			break Scan
+		}
+
+		var status string
+		if _, ferr := w.Write(nil); ferr != nil {
```

Comment:
Maybe add a comment here that `w.Write(nil)` actually writes a pkt-line flush packet (`0004`) as this is not obvious?

I wonder if `w` should be a `PacketWriter` instead of a `io.Writer`? Then we could use the actual `flush()` function. I feel that could be helpful for our future-selves reading this code 😄